storage module test fails when run all together

It seems we have some side effect between test cause when I run the full test suite if the storage module I got some errors:
```
--- FAIL: TestBtrfsRaid1 (1.40s)
    --- PASS: TestBtrfsRaid1/test_mounted (0.01s)
    --- PASS: TestBtrfsRaid1/test_path (0.00s)
    --- PASS: TestBtrfsRaid1/test_mount (0.28s)
    --- PASS: TestBtrfsRaid1/test_no_subvolumes (0.01s)
    --- PASS: TestBtrfsRaid1/test_create_volume (0.01s)
    --- PASS: TestBtrfsRaid1/test_list_volumes (0.01s)
    --- PASS: TestBtrfsRaid1/test_subvolume_list_no_subvolumes (0.00s)
    --- PASS: TestBtrfsRaid1/test_limit_subvolume (0.02s)
    --- PASS: TestBtrfsRaid1/test_remove_subvolume (0.01s)
    --- FAIL: TestBtrfsRaid1/add_device (0.01s)
        btrf_test.go:313: 
                Error Trace:    btrf_test.go:313
                Error:          Received unexpected error:
                                /dev/loop2 appears to contain an existing filesystem (btrfs).
                                ERROR: use the -f option to force overwrite of /dev/loop2
                Test:           TestBtrfsRaid1/add_device
        btrf_test.go:314: 
    --- FAIL: TestBtrfsRaid1/remove_device (0.01s)
        btrf_test.go:321: 
                Error Trace:    btrf_test.go:321
                Error:          Received unexpected error:
                                ERROR: error removing device '/dev/loop0': unable to go below two devices on raid1
                Test:           TestBtrfsRaid1/remove_device
        btrf_test.go:322: 
    --- PASS: TestBtrfsRaid1/remove_second_device (0.01s)
=== RUN   TestBtrfsList
--- FAIL: TestBtrfsList (0.65s)
    btrf_test.go:359: 
                Error Trace:    btrf_test.go:359
                Error:          Received unexpected error:
                                pool none has no corresponding devices on the system
                Test:           TestBtrfsList
--- FAIL: TestBtrfsRaid1 (1.40s)
    --- PASS: TestBtrfsRaid1/test_mounted (0.01s)
    --- PASS: TestBtrfsRaid1/test_path (0.00s)
    --- PASS: TestBtrfsRaid1/test_mount (0.28s)
    --- PASS: TestBtrfsRaid1/test_no_subvolumes (0.01s)
    --- PASS: TestBtrfsRaid1/test_create_volume (0.01s)
    --- PASS: TestBtrfsRaid1/test_list_volumes (0.01s)
    --- PASS: TestBtrfsRaid1/test_subvolume_list_no_subvolumes (0.00s)
    --- PASS: TestBtrfsRaid1/test_limit_subvolume (0.02s)
    --- PASS: TestBtrfsRaid1/test_remove_subvolume (0.01s)
    --- FAIL: TestBtrfsRaid1/add_device (0.01s)
        btrf_test.go:313: 
                Error Trace:    btrf_test.go:313
                Error:          Received unexpected error:
                                /dev/loop2 appears to contain an existing filesystem (btrfs).
                                ERROR: use the -f option to force overwrite of /dev/loop2
                Test:           TestBtrfsRaid1/add_device
        btrf_test.go:314: 
    --- FAIL: TestBtrfsRaid1/remove_device (0.01s)
        btrf_test.go:321: 
                Error Trace:    btrf_test.go:321
                Error:          Received unexpected error:
                                ERROR: error removing device '/dev/loop0': unable to go below two devices on raid1
                Test:           TestBtrfsRaid1/remove_device
        btrf_test.go:322: 
    --- PASS: TestBtrfsRaid1/remove_second_device (0.01s)
=== RUN   TestBtrfsList
--- FAIL: TestBtrfsList (0.65s)
    btrf_test.go:359: 
                Error Trace:    btrf_test.go:359
                Error:          Received unexpected error:
                                pool none has no corresponding devices on the system
                Test:           TestBtrfsList

```
While running each test one by one works without errors.

--- a/modules/go.mod
+++ b/modules/go.mod
@@ -52,9 +52,9 @@ require (
 	github.com/vishvananda/netns v0.0.0-20180720170159-13995c7128cc // indirect
 	github.com/vmihailenco/msgpack v4.0.4+incompatible // indirect
 	github.com/whs/nacl-sealed-box v0.0.0-20180930164530-92b9ba845d8d
-	golang.org/x/crypto v0.0.0-20190611184440-5c40567a22f8
-	golang.org/x/net v0.0.0-20190613194153-d28f0bde5980 // indirect
-	golang.org/x/sys v0.0.0-20190614084037-d442b75600c5
+	golang.org/x/crypto v0.0.0-20190701094942-4def268fd1a4
+	golang.org/x/net v0.0.0-20190724013045-ca1201d0de80 // indirect
+	golang.org/x/sys v0.0.0-20190804053845-51ab0e2deafa
 	golang.zx2c4.com/wireguard/wgctrl v0.0.0-20190607034155-226bf4e412cd
 	google.golang.org/appengine v1.6.1 // indirect
 	google.golang.org/genproto v0.0.0-20190611190212-a7e196e89fd3 // indirect

--- a/modules/storage/filesystem/btrf_test.go
+++ b/modules/storage/filesystem/btrf_test.go
@@ -31,10 +31,10 @@ import (
 
 type TestDevices map[string]string
 
-func (d TestDevices) Loops() []Device {
-	var loops []Device
+func (d TestDevices) Loops() DeviceCache {
+	var loops DeviceCache
 	for _, loop := range d {
-		loops = append(loops, Device{Path: loop})
+		loops = append(loops, &Device{Path: loop})
 	}
 
 	return loops
@@ -90,16 +90,13 @@ func setupDevices(count int) (devices TestDevices, err error) {
 }
 
 type TestDeviceManager struct {
-	devices TestDevices
+	devices DeviceCache
 }
 
-func (m TestDeviceManager) Device(ctx context.Context, device string) (*Device, error) {
+func (m TestDeviceManager) Device(ctx context.Context, path string) (*Device, error) {
 	for _, loop := range m.devices {
-		if loop == device {
-			return &Device{
-				Path: loop,
-				Type: "loop",
-			}, nil
+		if loop.Path == path {
+			return loop, nil
 		}
 	}
 
@@ -107,19 +104,17 @@ func (m TestDeviceManager) Device(ctx context.Context, device string) (*Device, 
 }
 
 func (m TestDeviceManager) ByLabel(ctx context.Context, label string) (DeviceCache, error) {
-	return nil, nil
+	var filterred DeviceCache
+	for _, device := range m.devices {
+		if device.Label == label {
+			filterred = append(filterred, device)
+		}
+	}
+	return filterred, nil
 }
 
 func (m TestDeviceManager) Devices(ctx context.Context) (DeviceCache, error) {
-	var devices DeviceCache
-	for _, loop := range m.devices {
-		devices = append(devices, &Device{
-			Path: loop,
-			Type: "loop",
-		})
-	}
-
-	return devices, nil
+	return m.devices, nil
 }
 
 func (m TestDeviceManager) Scan(ctx context.Context) error {
@@ -260,13 +255,10 @@ func TestBtrfsSingle(t *testing.T) {
 		t.Fatal("failed to initialize devices", err)
 	}
 	defer devices.Destroy()
+	loops := devices.Loops()
 
-	fs := NewBtrfs(TestDeviceManager{devices})
-	devs := DeviceCache{}
-	for idx := range devices.Loops() {
-		devs = append(devs, &devices.Loops()[idx])
-	}
-	pool, err := fs.Create(context.Background(), "test-single", devs, modules.Single)
+	fs := NewBtrfs(TestDeviceManager{loops})
+	pool, err := fs.Create(context.Background(), "test-single", loops, modules.Single)
 
 	if ok := assert.NoError(t, err); !ok {
 		t.Fatal()
@@ -284,12 +276,9 @@ func TestBtrfsRaid1(t *testing.T) {
 	defer devices.Destroy()
 
 	loops := devices.Loops()
-	fs := NewBtrfs(TestDeviceManager{devices})
-	devs := DeviceCache{}
-	for idx := range devices.Loops()[:2] {
-		devs = append(devs, &devices.Loops()[idx])
-	}
-	pool, err := fs.Create(context.Background(), "test-raid1", devs, modules.Raid1) //use the first 2 disks
+	fs := NewBtrfs(TestDeviceManager{loops})
+
+	pool, err := fs.Create(context.Background(), "test-raid1", loops[:2], modules.Raid1) //use the first 2 disks
 
 	if ok := assert.NoError(t, err); !ok {
 		t.Fatal()
@@ -309,7 +298,7 @@ func TestBtrfsRaid1(t *testing.T) {
 
 	t.Run("add device", func(t *testing.T) {
 		// add a device to array
-		err = pool.AddDevice(&loops[2])
+		err = pool.AddDevice(loops[2])
 		if ok := assert.NoError(t, err); !ok {
 			t.Fatal()
 		}
@@ -317,7 +306,7 @@ func TestBtrfsRaid1(t *testing.T) {
 
 	t.Run("remove device", func(t *testing.T) {
 		// remove device from array
-		err = pool.RemoveDevice(&loops[0])
+		err = pool.RemoveDevice(loops[0])
 		if ok := assert.NoError(t, err); !ok {
 			t.Fatal()
 		}
@@ -326,7 +315,7 @@ func TestBtrfsRaid1(t *testing.T) {
 	t.Run("remove second device", func(t *testing.T) {
 		// remove a 2nd device should fail because raid1 should
 		// have at least 2 devices
-		err = pool.RemoveDevice(&loops[1])
+		err = pool.RemoveDevice(loops[1])
 		if ok := assert.Error(t, err); !ok {
 			t.Fatal()
 		}
@@ -340,22 +329,23 @@ func TestBtrfsList(t *testing.T) {
 	}
 
 	defer devices.Destroy()
-
-	fs := NewBtrfs(TestDeviceManager{devices})
-
 	loops := devices.Loops()
+	fs := NewBtrfs(TestDeviceManager{loops})
+
 	names := make(map[string]struct{})
 	for i, loop := range loops {
 		name := fmt.Sprintf("test-list-%d", i)
 		names[name] = struct{}{}
-		_, err := fs.Create(context.Background(), name, DeviceCache{&loop}, modules.Single)
-
+		_, err := fs.Create(context.Background(), name, DeviceCache{loop}, modules.Single)
 		if ok := assert.NoError(t, err); !ok {
 			t.Fatal()
 		}
 	}
 
-	pools, err := fs.List(context.Background())
+	pools, err := fs.List(context.Background(), func(p Pool) bool {
+		return strings.HasPrefix(p.Name(), "test-")
+	})
+
 	if ok := assert.NoError(t, err); !ok {
 		t.Fatal()
 	}

--- a/modules/storage/filesystem/btrfs.go
+++ b/modules/storage/filesystem/btrfs.go
@@ -92,7 +92,10 @@ func (b *btrfs) Create(ctx context.Context, name string, devices DeviceCache, po
 	return newBtrfsPool(name, devices), nil
 }
 
-func (b *btrfs) List(ctx context.Context) ([]Pool, error) {
+func (b *btrfs) List(ctx context.Context, filter Filter) ([]Pool, error) {
+	if filter == nil {
+		filter = All
+	}
 	var pools []Pool
 	available, err := BtrfsList(ctx, "", false)
 	if err != nil {
@@ -110,12 +113,18 @@ func (b *btrfs) List(ctx context.Context) ([]Pool, error) {
 			return nil, err
 		}
 
+		pool := newBtrfsPool(fs.Label, devices)
+
+		if !filter(pool) {
+			continue
+		}
+
 		if len(devices) == 0 {
 			// since this should not be able to happen consider it an error
 			return nil, fmt.Errorf("pool %v has no corresponding devices on the system", fs.Label)
 		}
 
-		pools = append(pools, newBtrfsPool(fs.Label, devices))
+		pools = append(pools, pool)
 	}
 
 	return pools, nil

--- a/modules/storage/filesystem/device.go
+++ b/modules/storage/filesystem/device.go
@@ -74,9 +74,9 @@ func (l *lsblkDeviceManager) Devices(ctx context.Context) (DeviceCache, error) {
 func (l *lsblkDeviceManager) ByLabel(ctx context.Context, label string) (DeviceCache, error) {
 	var filtered DeviceCache
 
-	for idx := range l.devices {
-		if l.devices[idx].Label == label {
-			filtered = append(filtered, l.devices[idx])
+	for _, device := range l.devices {
+		if device.Label == label {
+			filtered = append(filtered, device)
 		}
 	}
 

--- a/modules/storage/filesystem/filesystem.go
+++ b/modules/storage/filesystem/filesystem.go
@@ -52,6 +52,14 @@ type Pool interface {
 	// Health() ?
 }
 
+// Filter closure for Filesystem list
+type Filter func(pool Pool) bool
+
+// All is default filter
+func All(Pool) bool {
+	return true
+}
+
 // Filesystem defines a filesystem interface
 type Filesystem interface {
 	// Create a new filesystem.
@@ -61,5 +69,5 @@ type Filesystem interface {
 	// profile: Raid profile of the filesystem
 	Create(ctx context.Context, name string, devices DeviceCache, profile modules.RaidProfile) (Pool, error)
 	// List all existing filesystems on the node
-	List(ctx context.Context) ([]Pool, error)
+	List(ctx context.Context, filter Filter) ([]Pool, error)
 }

--- a/modules/storage/storage.go
+++ b/modules/storage/storage.go
@@ -75,7 +75,7 @@ func (s *storageModule) initialize(policy modules.StoragePolicy) error {
 	// remount all existing pools
 	log.Info().Msgf("Remounting existing volumes")
 	log.Debug().Msgf("Searching for existing volumes")
-	existingPools, err := fs.List(ctx)
+	existingPools, err := fs.List(ctx, filesystem.All)
 	if err != nil {
 		return err
 	}

